### PR TITLE
Resource table tracks child relationships

### PR DESCRIPTION
**Pull request by pchickey**
First off, change all `wasmtime_wasi::preview2::TableError` cases to trap when converting to a filesystem::Error, streams::Error, etc. This is required by the shift to resources. No tests fail because of this change!

Introduce the tracking of children to `wasmtime_wasi::preview2::Table`. This required:
* a new pub method `push_child` indicating the parent
* Bookkeeping in a private `struct TableEntry { entry: Box<dyn Any...>, parent: Option<u32>, children: BTreeSet<u32> }`. This is the simplest datastructure that could possibly work.
* deletion checks that an entry has no children, and updates bookkeeping on the parent if the entry has one.
* re-defining `OccupiedEntry` to not use HashMap's directly, because we need to handle deletion bookkeeping.

There is only one child resource in wasmtime-wasi at the moment: a `HostPollable::TableEntry` is a pollable resource which refers to a parent with the `{ index: u32, ...}`. We changed `push_host_pollable` method on the table extension trait to push that variant with `push_child`.

I added a unit test to poll.rs showing that a `HostPollable::TableEntry` treats the given index as a parent.

I added an integration test to command-tests demonstrating correct management of child lifetimes exits correctly, and incorrect management traps with `TableError::HasChildren`.

This fixes a crash vector merged as part of https://github.com/bytecodealliance/wasmtime/pull/6556

### CODE REVIEW

File: crates/test-programs/build.rs [hunk 1]
```diff
@@ -172,7 +172,7 @@ fn components_rs(
     let mut decls = String::new();
     let mut cases = String::new();
     let mut uses = String::new();
-    for target_name in targets_in_package(&meta, package, kind) {
+    for target_name in dbg!(targets_in_package(&meta, package, kind)) {
```
Comment (elliottt) [inline crates/test-programs/build.rs:175]:
```suggestion
    for target_name in targets_in_package(&meta, package, kind) {
```

File: crates/wasi/src/preview2/table.rs [hunk 2]
```diff
@@ -20,21 +22,90 @@ pub enum TableError {
...
-    /// Create an empty table. New insertions will begin at 3, above stdio.
+    /// Create an empty table
     pub fn new() -> Self {
         Table {
             map: HashMap::new(),
-            next_key: 3, // 0, 1 and 2 are reserved for stdio
+            // 0, 1 and 2 are formerly (preview 1) for stdio. To prevent users from assuming these
+            // indicies are still valid ways to access stdio, they are deliberately left empty.
+            // Once we have a full implementation of resources, this confusion should hopefully be
+            // impossible :)
```
Comment (elliottt) [inline crates/wasi/src/preview2/table.rs:81]: Great comment!

**Review by elliottt — APPROVED** [review]
Seems reasonable to me!

File: crates/test-programs/build.rs (hunk 1 above)
Comment (pchickey) [reply crates/test-programs/build.rs:175]: oh gosh thank you

**Pull request merged**